File: src/types/sale.ts
import type { SaleDetail } from "./sale_detail";

export interface Sale {
  id: number;
  user_id: number;
  cashier_name: string;
  cashier_email: string;
  checkout_token: string;
  cart_fingerprint: string;
  sale_date: string;
  total: number;
  payment_method: PaymentMethod;
  cash_received: number | null;
  change_amount: number | null;
  created_at: string;
  updated_at: string;
}

export type PaymentMethod = "efectivo" | "tarjeta" | "transferencia";

export interface SaleReceipt {
  sale: Sale;
  details: SaleDetail[];
}

export interface SalesViewFilters {
  search?: string;
  payment_method?: PaymentMethod;
}

export interface SalesViewRow {
  id: number;
  ticket_number: string;
  sale_date: string;
  payment_method: PaymentMethod;
  total: number;
  item_count: number;
  user_id: number;
  user_name: string;
}

export interface SalesViewStats {
  sale_count: number;
  total_sold: number;
  average_sale: number;
  items_sold: number;
}

export interface SalesView {
  items: SalesViewRow[];
  stats: SalesViewStats;
}

export interface SaleHistoryLine {
  product_id: number;
  product_name: string;
  sku: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
}

export interface SaleHistoryDetail {
  id: number;
  ticket_number: string;
  payment_method: PaymentMethod;
  sale_date: string;
  user_id: number;
  user_name: string;
  user_email: string;
  lines: SaleHistoryLine[];
  item_count: number;
  total: number;
  cash_received: number | null;
  change_amount: number | null;
}

File: src-tauri/src/models/sale.rs
use serde::{Deserialize, Serialize};

use super::sale_detail::SaleDetail;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Sale {
    pub id: i64,
    pub user_id: i64,
    pub cashier_name: String,
    pub cashier_email: String,
    pub checkout_token: String,
    pub cart_fingerprint: String,
    pub sale_date: String,
    pub total: i64,
    pub payment_method: String,
    pub cash_received: Option<i64>,
    pub change_amount: Option<i64>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct SaleReceipt {
    pub sale: Sale,
    pub details: Vec<SaleDetail>,
}

#[derive(Debug, Default, Deserialize)]
pub struct SalesViewFilters {
    pub search: Option<String>,
    pub payment_method: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SalesViewRow {
    pub id: i64,
    pub ticket_number: String,
    pub sale_date: String,
    pub payment_method: String,
    pub total: i64,
    pub item_count: i64,
    pub user_id: i64,
    pub user_name: String,
}

#[derive(Debug, Serialize)]
pub struct SalesViewStats {
    pub sale_count: i64,
    pub total_sold: i64,
    pub average_sale: i64,
    pub items_sold: i64,
}

#[derive(Debug, Serialize)]
pub struct SalesView {
    pub items: Vec<SalesViewRow>,
    pub stats: SalesViewStats,
}

#[derive(Debug, Serialize)]
pub struct SaleHistoryLine {
    pub product_id: i64,
    pub product_name: String,
    pub sku: String,
    pub quantity: i64,
    pub unit_price: i64,
    pub subtotal: i64,
}

#[derive(Debug, Serialize)]
pub struct SaleHistoryDetail {
    pub id: i64,
    pub ticket_number: String,
    pub payment_method: String,
    pub sale_date: String,
    pub user_id: i64,
    pub user_name: String,
    pub user_email: String,
    pub lines: Vec<SaleHistoryLine>,
    pub item_count: i64,
    pub total: i64,
    pub cash_received: Option<i64>,
    pub change_amount: Option<i64>,
}
